### PR TITLE
Pin numba version

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Changes in the upcoming numba 0.57 release will break cudf until we've had time to adapt them, which will not happen in 23.04.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

File: dependencies.yaml [hunk 1]
```diff
@@ -340,7 +340,7 @@ dependencies:
         packages:
           - cachetools
           - cuda-python>=11.7.1,<12.0
-          - numba>=0.56.2
+          - &numba numba>=0.56.2,<0.57
```
[the remark this answers is not in the document]
Comment (bdice) [reply dependencies.yaml:343]: YAML anchor. The & defines it and the * reuses that content.

**Review by gmarkall — COMMENTED** [review]
There were some CUDA-related fixes for extensions that went into Numba 0.56.4 - while we're here could / should we bump it to `>=0.56.4,<0.57`?

FWIW, `<0.57` looks like the correct upper bound to me from a Numba standpoint but I can't vouch for the changes (whether there's other locations that need changing in cudf).

File: dependencies.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (galipremsagar) [reply dependencies.yaml:343]: I see, TIL. Thanks 👍

**Review by galipremsagar — APPROVED** [review]

**Comment by bdice** [comment]
@gmarkall @brandon-b-miller @vyasr Can we cross-link an issue describing the problem with Record types / UDFs? Anything to document why we’re pinning in a little more detail, so that we can resolve it later when 0.57 is out and have everything link together.

**Comment by gmarkall** [comment]
@bdice Created in #13002 to record the general idea - I don't have a working cuDF environment right now so the issue doesn't have e.g. the traceback and a complete environment description attached, but the problem is well-understood.

File: dependencies.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply dependencies.yaml:343]: This creates a [YAML anchor](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/) that allows it to be referenced later (see the other place I'm doing `*numba`)

File: dependencies.yaml [hunk 2]
```diff
@@ -340,7 +340,7 @@ dependencies:
         packages:
           - cachetools
           - cuda-python>=11.7.1,<12.0
-          - numba>=0.56.2
+          - &numba numba>=0.56.4,<0.57
```
Comment (gmarkall) [inline dependencies.yaml:343]: Probably a basic question, but why does this change into `& numba numba`?

File: dependencies.yaml [hunk 3]
```diff
@@ -493,4 +493,4 @@ dependencies:
       - output_types: [conda, requirements, pyproject]
         packages:
           - dask-cuda==23.4.*
-          - numba>=0.56.2
+          - *numba
```
Comment (gmarkall) [inline dependencies.yaml:496]: Is this related to the `&numba numba` above?

**Review by gmarkall — APPROVED** [review]
LGTM (from the perspective of the Numba version ranges).

File: dependencies.yaml (hunk 3 above)
Comment (wence-) [reply dependencies.yaml:496]: Yes, the `&numba ...` takes a reference to `...` and "stores" it in `numba` which can then be dereferenced and copied in with `*numba`.

**Review by wence- — APPROVED** [review]

File: dependencies.yaml (hunk 2 above)
Comment (ajschmidt8) [reply dependencies.yaml:343]: This is referred to as a YAML anchor. You can view the spec here: https://yaml.org/spec/1.2.2/#692-node-anchors.

There are some Google results that provide better examples than the official spec docs above.

In this case an anchor was used to keep the `dependencies.yaml` file DRY.

**Review by ajschmidt8 — APPROVED** [review]
